// eslint-import-resolver-parcel/dist-node/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var path = require('path');
var fs = _interopDefault(require('fs'));
var resolve$1 = require('resolve');

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _objectSpread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};
    var ownKeys = Object.keys(source);

    if (typeof Object.getOwnPropertySymbols === 'function') {
      ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) {
        return Object.getOwnPropertyDescriptor(source, sym).enumerable;
      }));
    }

    ownKeys.forEach(function (key) {
      _defineProperty(target, key, source[key]);
    });
  }

  return target;
}

const defaultExtensions = ['.js', '.jsx'];
const interfaceVersion = 2;

function findAliases() {
  let currentPath = path.posix.resolve();
  let packagePath = path.posix.resolve(currentPath, 'package.json');

  while (!fs.existsSync(packagePath)) {
    currentPath = path.posix.resolve(currentPath, '..');
    packagePath = path.posix.resolve(currentPath, 'package.json');
  }

  const packageJson = fs.readFileSync(packagePath);
  return JSON.parse(packageJson).alias || {};
}

function resolvePackageLevel(source, file, rootDirectory) {
  let packageDirectory = path.posix.dirname(file);

  for (;;) {
    if (fs.existsSync(path.posix.resolve(packageDirectory, 'node_modules'))) {
      // found node_modules
      break;
    } else if (path.posix.parse(packageDirectory).root === packageDirectory) {
      // reached the drive root, just return null
      return null;
    }

    packageDirectory = path.posix.resolve(packageDirectory, '..');
  }

  return path.posix.join( // return rootDirectory if it is nested inside the packageDirectory
  rootDirectory.startsWith(packageDirectory) ? rootDirectory : packageDirectory, // Get rid of the tilde
  source.slice(1));
}

function resolve(source, file, possibleConfig = {}) {
  if (resolve$1.isCore(source)) return {
    found: true,
    path: null
  };

  const config = _objectSpread({
    rootDir: '',
    extensions: []
  }, possibleConfig);

  const [startSource] = source.split('/');
  const aliases = findAliases();
  const foundAlias = Object.keys(aliases).find(alias => alias === startSource);
  let newSource = source.replace(foundAlias, aliases[foundAlias]);
  const rootDirectory = path.posix.resolve(config.rootDir);

  switch (newSource[0]) {
    case '.':
      newSource = path.posix.resolve(path.posix.dirname(file), newSource);
      break;

    case '~':
      newSource = resolvePackageLevel(newSource, file, rootDirectory);
      break;

    case '/':
      newSource = path.posix.join(rootDirectory, newSource);
      break;
    // no default
  }

  try {
    return {
      found: true,
      path: resolve$1.sync(newSource, {
        basedir: path.posix.resolve(),
        extensions: [...defaultExtensions, ...config.extensions]
      })
    };
  } catch (_) {
    return {
      found: false
    };
  }
}

exports.interfaceVersion = interfaceVersion;
exports.resolve = resolve;
